src/Orleans.Core/Providers/LegacyProviderConfigurator.cs

```diff
@@ -106,31 +108,43 @@ public ProviderLifecycleParticipant(IProviderConfiguration config, IServiceProvi
                 this.defaultInitStage = defaultInitStage;
                 this.defaultStartStage = defaultStartStage;
                 this.schedule = services.GetService<LegacyProviderConfigurator.ScheduleTask>();
+                this.provider = new Lazy<IProvider>(() => services.GetServiceByName<TService>(this.config.Name) as IProvider);
             }
 
             public virtual void Participate(TLifecycle lifecycle)
             {
                 this.initStage = this.config.GetIntProperty(LegacyProviderConfigurator.InitStageName, this.defaultInitStage);
-                lifecycle.Subscribe(this.initStage, this.Init, this.Close);
+                lifecycle.Subscribe(this.initStage, this.Init, this.ProviderClose);
                 this.startStage = this.config.GetIntProperty(LegacyProviderConfigurator.StartStageName, this.defaultStartStage);
-                lifecycle.Subscribe(this.startStage, this.Start);
+                lifecycle.Subscribe(this.startStage, this.Start, this.StreamProviderClose);
 
             }
 
             private async Task Init(CancellationToken ct)
             {
                 var stopWatch = Stopwatch.StartNew();
-                IProvider provider = this.services.GetServiceByName<TService>(this.config.Name) as IProvider;
+                IProvider provider = this.provider.Value;
```

Comment:
what's the reason of lazy initialization?

Reply:
Can't resolve concrete types at construction time because parts loader may not have loaded assemblies yet.